components/persistor/lib/knex/db.ts

```diff
@@ -805,41 +820,52 @@ module.exports = function (PersistObjectTemplate) {
 
                     name = name.toLowerCase();
                     if (operation === 'add') {
-                        table[type](columns, name);
+                        await knex.schema.table(tableName, function (table) {
+                            table[type](columns, name);
+                        }).catch((error) => {
+                            logTableChangesWarningMessage(columns, type, name, error);
+                        })
                     }
                     else if (operation === 'delete') {
                         type = type.replace(/index/, 'Index');
                         type = type.replace(/unique/, 'Unique');
-                        table['drop' + type]([], name);
-                    }
-                    else
-                        table[type](columns, name);
-
-                }));
-            }
-
-
-            return knex.transaction(function (trx) {
-                return trx.schema.table(tableName, function (table) {
-                    syncIndexesForHierarchy('delete', dbChanges, table);
-                    syncIndexesForHierarchy('add', dbChanges, table);
-                    syncIndexesForHierarchy('change', dbChanges, table);
-                }).catch(function (error) {
-                    if (error && error.message && error.message === 'index type can be only "unique" or "index"') {
-                        throw error;
+                        await knex.schema.table(tableName, function (table) {
+                            table['drop' + type]([], name);
+                        }).catch((error) => {
+                            logTableChangesWarningMessage(columns, type, name, error);
+                        });
                     }
-                    const logger = PersistObjectTemplate && PersistObjectTemplate.logger;
-                    if (logger) {
-                        logger.warn({
-                            component: 'persistor',
-                            dbChanges: dbChanges,
-                            module: 'db',
-                            activity: 'synchronizeIndexes',
-                            error: error
-                        }, 'Unable to update index');
+                    else {
+                        await knex.schema.table(tableName, function (table) {
+                            table[type](columns, name);
+                        }).catch((error) => {
+                            logTableChangesWarningMessage(columns, type, name, error);
+                        });
+                        
+                        }
                     }
-                });
-            });       
+                ));
+            }
+            try { 
+                await syncIndexesForHierarchy('delete', dbChanges);
+                await syncIndexesForHierarchy('add', dbChanges);
```

Comment:
is having all 3 of the calls to `syncIndexesForHierarchy` in one try catch ok ?

Reply:
I have switched things around. Old commit. Let me know if what I have now makes sense!